### PR TITLE
Add a fuzzy/regex pattern-matching for metric allow and block list

### PR DESCRIPTION
**Pull request by ferruzzi**
Current metric allow and block lists allow the user to set which metric names to emit, but they only match on `startswith()`.  This PR adds two new Validators - FuzzyAllow and FuzzyBlock, which match using regex - along with a config option to chooze between the "basic" or "fuzzy" match and unit testing.

While I was in there I also noticed that the OTel metrics never implemented Blocklist, it only had an AllowList.  This could maybe have been a separate PR, but I fixed that bug while I was working on it since that would have required adjustments either way.

The new `get_validator()` method is a modified version of the decision logic previously used in the statsd logger, expanded to include the Fuzzy options, and then also applied to the otel logger.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/config_templates/config.yml
Comment (shubham22) [inline airflow/config_templates/config.yml]: I am personally against adding a new config for `metrics_use_fuzzy_match`. With this approach, we will keep sprawling the number of configurations every time we need to make a hard decision. I am of the opinion that we should have just 2 configs - `metrics_allow_list` and `metrics_block_list` - and they should be changed to support the pattern matching. That means, we should deprecate existing configs now and introduce new ones which are more powerful and support pattern matching, in order to keep it backward compatible. Yes, this means we need to go through deprecation until v3.0, but I think that is better for the product.

**Review by shubham22 — COMMENTED** [review]
This is great @ferruzzi. Giving more control to the users would help them get the right set of metrics in their monitoring tools, without spending on other ones.

File: airflow/config_templates/config.yml
Comment (shubham22) [reply airflow/config_templates/config.yml]: I would also suggest renaming this to `metrics_enable_pattern_match` or `metrics_use_pattern_match`, in case we really decide to go with this.

Comment (hussein-awala) [reply airflow/config_templates/config.yml]: I like the idea of pattern match, we even don't need to rename the variable, we can just update the configuration description and explain that since 2.9.0, Airflow supports regex in `metrics_allow_list` and `metrics_block_list`.

Comment (ferruzzi) [reply airflow/config_templates/config.yml]: @hussein-awala   This feels like a breaking change to me, no?  What previously only matched the start of the metric name string will now match any substring in the name.

Comment (ferruzzi) [reply airflow/config_templates/config.yml]: @shubham22 - I'm not against the idea of deprecating but there needs to be some kind of extra config option in the interim. If we are going to deprecate then our best option (as I see it) is to keep `metric_allow_list` and `metric_block_list`, add a flag to pick "old style or new style", and add a deprecation message if that is set to "old style".  Which is what I've done, minus the message.   I can add the message if we don't want to have the option to keep the old way, but we're going to be stuck with this extra config option until (3.0??) and then likely a bunch of confusion when the default changes after that.

Comment (shubham22) [reply airflow/config_templates/config.yml]: > then likely a bunch of confusion when the default changes after that.

This is the primary reason I think moving over to two new configs that replace old configs as better. This way when the old config is deprecated, users who adopted the new configs wouldn't have to make any change at all.

Comment (o-nikolas) [reply airflow/config_templates/config.yml]: >This is the primary reason I think moving over to two new configs that replace old configs as better. This way when the old config is deprecated, users who adopted the new configs wouldn't have to make any change at all.

But now isn't this adding even more new configs than Dennis's original proposal?

Ultimately, I agree that we should change the definition of the current configs and add nothing new (zero new configs added) but we also do need _some_ way to give people fair warning. I think we should try do that with adding as few, if any, configs as possible. So I'm in slight favour of Dennis's proposal since it only adds one new config, which will be dropped after the deprecation process/switch over. Another approach could be a version check in the code that says something like "the behaviour of these configs will change by release <foo>, please be prepared to update your configuration at that time" (or a better message since I just spit balled that right now...) and uses the old beahviour if the code is before the release where we switch, and later releases it uses the new behaviour.

Comment (ferruzzi) [reply airflow/config_templates/config.yml]: It looks like there are a number of "less-bad" solutions but no real good one.   Does anyone feel particularly strongly about any of the options here?

If nobody has a strong opinion, I'm going to add "if metrics_use_fuzzy_match is false, display a deprecation notice" and leave it at that for now.

Comment (o-nikolas) [reply airflow/config_templates/config.yml]: I don't feel _too_ strongly, not willing to die on any hills here.

**Comment by ferruzzi** [comment]
Alright, based on feedback above, I am going to add a warning (once on launch) if `metrics_use_fuzzy_match` is false stating that the basic matching will be deprecated in the future and you can try the new pattern matching by setting that config option to True.

I also thought of two more unit tests which confirm the allow list can handle regex (both a pass and a fail case) and not just simple substring matching.

Based on the above feedback, I'm also renaming from "fuzzy matching" to "pattern matching".

**Comment by ferruzzi** [comment]
Slight git goof, sorry for the rebase and force push.

File: airflow/config_templates/config.yml
Comment (ferruzzi) [reply airflow/config_templates/config.yml]: Based on discussion here, I'm going to add a deprecation warning and rename "fuzzy matching" to "pattern matching".

File: tests/core/test_stats.py
Comment (o-nikolas) [inline tests/core/test_stats.py]: Slight nit: all these patterns match from the beginning of the string either by forcing it with ^ or just by chance (with the middle one). Since we want to ensure this new feature can match in the middle, do you mind adding a regex that does so? I think you could actually just make the 6 a 5 and it would match in the middle.

File: airflow/config_templates/config.yml
Comment (o-nikolas) [inline airflow/config_templates/config.yml]: A suggestion to make this even more concrete and spelled out. Seems obvious to us working on it, but it's a slightly tricky one if you aren't familiar.
```suggestion
        If true, metrics_allow_list and metrics_block_list will use regex pattern matching anywhere within the metric
        name instead of only prefix matching at the start of the name.
```

**Review by o-nikolas — APPROVED** [review]
Left a couple nit comments. Otherwise I think this is good to go :+1:

File: tests/core/test_stats.py
Comment (ferruzzi) [reply tests/core/test_stats.py]: Gah, you are absolutely right.  I intended to test that in the middle one and goofed.  I'll adjust it to `a.{4}o` which verifies pattern matching (and not just substring) and matches not-the-beginning

File: airflow/config_templates/config.yml
Comment (shubham22) [inline airflow/config_templates/config.yml]: change the version added, once the milestone is updated. If possible, let's do it in 2.8.1

File: airflow/config_templates/config.yml
Comment (shubham22) [inline airflow/config_templates/config.yml:949]: I would suggest changing description of "metrics_allow_list" and/or "metrics_block_list" to mention about behaviour change when this is "True"

File: tests/core/test_stats.py
Comment (shubham22) [inline tests/core/test_stats.py:343]: to have more robust test, shouldn't you actually use regex when pattern match is on? This test will pass irrespective of pattern match configuration value.

**Review by shubham22 — APPROVED** [review]

File: tests/core/test_stats.py
Comment (ferruzzi) [reply tests/core/test_stats.py:343]: There is a whole separate group of tests using regex patterns.  This shows that you can just enter a string and it'll match anywhere in the name, and also optionally you can use regex patterns for the more complex stuff if you wish.

File: airflow/config_templates/config.yml
Comment (ferruzzi) [reply airflow/config_templates/config.yml]: Done, thanks

File: airflow/config_templates/config.yml
Comment (ferruzzi) [reply airflow/config_templates/config.yml:949]: I'll workshop a couple of ideas.  Thanks for the suggestion

File: airflow/config_templates/config.yml
Comment (hussein-awala) [inline airflow/config_templates/config.yml]: Since it's a new feature, we need to wait for the next minor release:
```suggestion
      version_added: 2.9.0
```

File: airflow/metrics/validators.py
Comment (hussein-awala) [inline airflow/metrics/validators.py]: This is a part of Airflow core, you cannot use `AirflowProviderDeprecationWarning` warning:
```suggestion
            RemovedInAirflow3Warning,
```

**Review by hussein-awala — COMMENTED** [review]
Looks good, I just added two small comments, LGTM when you fix them.

File: airflow/config_templates/config.yml
Comment (ferruzzi) [reply airflow/config_templates/config.yml:949]: Revised

**Review by hussein-awala — APPROVED** [review]

**Pull request merged**
